File: Assets/Greco3D/scripts/task/Timer.js
#pragma strict

private var stopwatch:float;
private var trial_action : String = "run_start";
private var total_time : float;
public var timer : float = 0;
public var cnt_trial : int;
public var iti_time : float;
public var trial_time : float;
public var numTrials : int;

function GetTime(type : String):float{
	var getTime : float;
	switch(type){
		case 'task':
			getTime = stopwatch;		
			break;
		case 'trial':
			getTime = timer;
			break;
	}
	return timer;
}

function GetAction():String{

	return trial_action;
}

function SetUpTime(iti : float, trial : float, numT : int){
		iti_time = iti;
		trial_time = trial;
		total_time = iti + trial;
		timer = total_time;
		stopwatch = 0;
		trial_action = "run_start";
		numTrials = numT;
		cnt_trial = 0;
		yield;
}

function StartRunClock(){
	while(true){
	stopwatch += Time.deltaTime;
	}
	yield;
}

function StartTimerITI(){
	timer = 0;
	while(timer < iti_time){
	timer += Time.deltaTime;
	yield;
	}
	timer = 0;
}

function StartTimerTrial(){
	timer = 0;
	while(timer < trial_time){
	timer += Time.deltaTime;
	yield;
	}
	timer = 0;
}

function TrialTimer(){
print("start trial timer");
trial_action = "trial";
yield StartCoroutine(StartTimerTrial());
print("start iti timer");
trial_action = "iti";
yield StartCoroutine(StartTimerITI());
cnt_trial++;
}

function StartRun(){
print("start iti timer");
trial_action = "iti";
yield StartCoroutine(StartTimerITI());
while(cnt_trial<numTrials){
yield StartCoroutine(TrialTimer());
trial_action = "run_end";
}
}

File: Assets/Greco3D/Tasks/Timer.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class Timer : MonoBehaviour {
    public string trialAction;
    public float stopWatch, timer,totalTime, itiTime, trialTime;
    public int cntTrial, numTrials;

    public float GetTime(string type)
    {
        float getTime = 0f;
        switch (type)
        {
         case "task":
            getTime = stopWatch;
            break;
        case "trial":
            getTime = timer;
            break;
        }
        return getTime;
    }

    public string GetAction()
    {
        return trialAction;
    }

    public IEnumerator SetUpTime(float iti, float trial, int numT)
    {
        itiTime = iti;
        trialTime = trial;
        totalTime = iti + trial;
        timer = totalTime;
        stopWatch = 0f;
        trialAction = "run_start";
        numTrials = numT;
        cntTrial = 0;
        yield return null;
    }

    IEnumerator StartRunClock()
    {
        while (true)
        {
            stopWatch += Time.deltaTime;
            yield return null;
        }
    }

    IEnumerator StartTimerITI()
    {
        timer = 0f;
        while (timer < itiTime)
        {
            timer += Time.deltaTime;
            yield return null;
        }
    }

    IEnumerator StartTimer(float time)
    {
        timer = 0f;
        while (timer < time)
        {
            timer += Time.deltaTime;
            yield return null;
        }
    }

    IEnumerator StartTimerTrial()
    {
        timer = 0f;
        while (timer < trialTime)
        {
            timer += Time.deltaTime;
            yield return null;
        }
    }

    IEnumerator TrialTimer()
    {
        trialAction = "trial";
        yield return StartCoroutine(StartTimerTrial());
       
        trialAction = "iti";
        cntTrial++;
        yield return StartCoroutine(StartTimerITI());
    }

    public IEnumerator StartRun()
    {
        trialAction = "start";
        StartCoroutine(StartRunClock());
        yield return StartCoroutine(StartTimer(2f));
        do
        {
            yield return StartCoroutine(TrialTimer());
        } while (cntTrial < numTrials);
        StopCoroutine(StartRunClock());
        trialAction = "run_end";
    }
}
